File: Leetcode/HumbleNumbers/03_longest-substring-without-repeating-characters/03longest-substring-without-repeating-characters.py
from codecs import lookup


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        length=len(s)
        current_length=0
        max_length=0
        lookup=set()
        left=0
        for i in range(length):
            current_length+=1
            while s[i] in lookup:
                lookup.remove(s[left])
                current_length-=1
                left+=1
            if current_length > max_length:max_length = current_length
           
            lookup.add(s[i])
        return max_length
def main():
    k=Solution()
    s="abcabcbb"
    t=k.lengthOfLongestSubstring(s)
    print("t="+str(t))
if __name__=="__main__":
    main()
                
       

File: Leetcode/HumbleNumbers/03_longest-substring-without-repeating-characters/C++/03longest-substring-without-repeating-characters.cpp
#include <iostream>
#include <unordered_map>  //map
#include <string> //string
using namespace std;
class solution {
  public:  int lengthOfLongestSubstring(string s)
    {
        unordered_map<char,int>windows;
        int left=0;
        int length=s.length();
        int max_length=0;
        for(int i=0;i<s.length();i++)
        {
            windows[s[i]]+=1;
            // 左窗口收缩，只有右边届的引入，才可能导致字符重复
            while(windows[s[i]]==2)
            {
                windows[s[left]]-=1;
                left++;
            }
           max_length=max(max_length,i-left+1);
        }
        return max_length;
    }
};
int main()
{
    
    string s1="abcabcbb";
    string s2="pwwkew";
    string s3="bbbbbbb";
    solution s;
   int a= s.lengthOfLongestSubstring(s1);
   int b= s.lengthOfLongestSubstring(s2);
   int c= s.lengthOfLongestSubstring(s3);
   cout<<"a="<<a<<endl;
   cout<<"b="<<b<<endl;
   cout<<"c="<<c<<endl;
}